lib/chef/provider/user/mac.rb

```diff
@@ -0,0 +1,607 @@
+          #
+          # NOTE: `dscl -read` and `plutil -convert` return different values for ShadowHashData.
+          #
+          # `dscl` returns the value encoded as an octal hex string and stored as a <string>
+          # `plutil` returns the value encoded as a base64 string stored as <data>
+          #
+          #  eg:
+          #
+          # <array>
+          #   <string>77687920 63616e27 74206170 706c6520 6275696c 6420636f 6e736973 74656e74 20746f6f 6c696e67</string>
+          # </array>
+          #
+          # vs
+          #
+          # <array>
+          #   <data>AADKAAAKAA4LAA0MAAAAAAAAAAA=</data>
+          # </array>
+          #
+          shadow_hash_hex = dscl_get(user_plist, :shadow_hash)[0]
+
+          return user_plist unless shadow_hash_hex && shadow_hash_hex != ""
+
+          shadow_binary_plist = [shadow_hash_hex.delete(" ")].pack("H*")
+          shadow_xml_plist = shell_out("plutil", "-convert", "xml1", "-o", "-", "-", input: shadow_binary_plist).stdout
+          dscl_set(user_plist, :shadow_hash, Plist.parse_xml(shadow_xml_plist))
+
+          user_plist
+        rescue Chef::Exceptions::PlistUtilCommandFailed, Chef::Exceptions::DsclCommandFailed
+          nil
+        end
+
+        # Map updated user information from the user_plist onto the current resource
+        # and instance variables.
+        def reload_with_user_plist
+          current_resource.uid(dscl_get(user_plist, :uid)[0])
+          current_resource.gid(dscl_get(user_plist, :gid)[0])
+          current_resource.home(dscl_get(user_plist, :home)[0])
+          current_resource.shell(dscl_get(user_plist, :shell)[0])
+          current_resource.comment(dscl_get(user_plist, :comment)[0])
+
+          shadow_hash = dscl_get(user_plist, :shadow_hash)
+          if shadow_hash
+            current_resource.password(shadow_hash[0]["SALTED-SHA512-PBKDF2"]["entropy"].string.unpack("H*")[0])
+            current_resource.salt(shadow_hash[0]["SALTED-SHA512-PBKDF2"]["salt"].string.unpack("H*")[0])
+            current_resource.iterations(shadow_hash[0]["SALTED-SHA512-PBKDF2"]["iterations"].to_i)
+          end
+
+          @auth_authority = dscl_get(user_plist, :auth_authority)
+          @guid = dscl_get(user_plist, :guid)
+          @admin_group_info = nil
+
+          current_resource.secure_token(secure_token_enabled?)
+          current_resource.admin(admin_user?)
+        end
+
+        #
+        # User Provider Callbacks
+        #
+
+        def create_user
+          cmd = [-"-addUser", new_resource.username]
+          cmd += ["-fullName", new_resource.comment]  if new_resource.property_is_set?(:comment)
+          cmd += ["-UID", new_resource.uid]           if new_resource.property_is_set?(:uid)
+          cmd += ["-shell", new_resource.shell]       if new_resource.property_is_set?(:shell)
+          cmd += ["-home", new_resource.home]         if new_resource.property_is_set?(:home)
```

Comment:
for some of these like home directory you set a default and don't allow setting it to nil, so the `if new_resource.property_is_set?(:home)` check can likely be omitted, because it will always have a valid setting that should be respected.

but even for properties that could be nil you should avoid the `property_is_set?` check and just use `if new_resource.whatever` or `unless new_resource.whatever.nil?` (if its boolean-valued).

Reply:
(generally you probably don't want to be using `property_is_set?` at all because like i said offhand i can't recall if a default value is 'set' or not and if i have to read the code to determine how it is implemented and how close to `nil?` the check is or not then you're likely asking too much of the reader in most cases)

Reply:
yeah i'm a bit confused, i was gonna say that this should be `diverged?(:home)` and push the `property_is_set?` consistently into that API, but this is creating where you don't want to worry about "is it diverged, and did the user want to fix it?" but instead you just want to create the user, so I think you want to just always set the home here -- otherwise you won't be calling with the default you specified in the resource (which might work but smells buggy one way or another).